NVDA menu: add explicit shortcut keys and avoid collisions

CyrilleB79 opened the pull request:
### Link to issue number:
Closes #15362

### Summary of the issue:
"Add-on store" item has been added with "S" as a shortcut key in NVDA tools menu in NVDA 2023.2.
Unfortunately "S" was already used for "Speech viewer" (as its first letter), even if it was not defined explicitly. Now that there is an explicit mapping for "Add-on store", implicit shortcut key definition for "Speech viewer" becomes masked and does not work anymore.

In case the add-ons add new items in the menus with explicit definition of shortcut keys, other menu item that only use implicit definition of the shortcut key may be masked as well.

### Description of user facing changes
Explicitly define shortcut keys in main NVDA menu and avoid collisions (i.e. two times the same letter in the same menu).

### Description of development approach
In main NVDA menu:
* "Tools: add explicitly "T"
* "Reset configuration to factory defaults")": Change "R" to "F" because was already used by "Revert to saved configuration"
* "Donate": add explicitly "D"

In "Tools" submenu:
* "Reload plugins e": change "R" to "E" since "R" already taken by "Run COM Registration Fixing tool..."
* "Add-on store": "A" instead of "S", as discussed in https://github.com/nvaccess/nvda/issues/15362#issuecomment-1704877030, to avoid collision with "Speech viewer"
* "View log": "L" instead of "V" to be more intuitive, as discussed in https://github.com/nvaccess/nvda/pull/15364#discussion_r1315305640
* For all other items, explicitly mark the accelerator

Help menu:
* "About": Explicitly set "A"
* All other items are already explicitly marked.

### Testing strategy:

* Visually check that all items are explicitly defined, i.e. that the shortcut key is underlined; an implicit definition targetting the first letter would not show an underlined character instead.
* Try to activate all the items that have been changed with their shortcut key

### Known issues with pull request:
There may still be shortcut collisions or masked implicit definitions in this menu for items added by add-ons. It's up to add-on author to adapt if needed.
### Change log entries:
Not needed
### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

CyrilleB79 committed:
NVDA menu: add explicit shortcut keys and avoid collisions

CyrilleB79 committed:
fix linting error

seanbudd commented on source/gui/__init__.py:
```diff
@@ -494,17 +494,17 @@ def __init__(self, frame: MainFrame):
 		if not globalVars.appArgs.secure:
 			# Translators: The label for the menu item to open NVDA Log Viewer.
-			item = menu_tools.Append(wx.ID_ANY, _("View log"))
+			item = menu_tools.Append(wx.ID_ANY, _("&View log"))
```
Perhaps this should be changed to `&l`? It's a bit more intuitive

seanbudd replied:
```suggestion
			item = menu_tools.Append(wx.ID_ANY, _("View &log"))
```

CyrilleB79 replied:
In my idea, I have only made explicit shortcuts tha were previously implicit when possible. Since "V" was used previously as an implicit shortcut, I have kept it.

Your suggestion makes sense, but I do not know if there are training material (from NV Access or others) which indicate NVDA+N, T, V to open the log viewer. I would like to have @Qchristensen's opinion about it before accepting the suggestion.

Qchristensen replied:
I don't have a strong preference either way here - I don't think I've written any instructions (not recently at least, and not in say Basic Training for NVDA - I mention the log viewer and where it is, but don't go through the specific keys to press to open it).  If there are commonly used community instructions which mention it, as @XLTechie notes is the case with the COM Registration fixing tool in the conversation further down, then I'd be happy to follow their lead, but otherwise, I'm happy either way here - there is some logic to L, given the first three options (view log, braille viewer and speech viewer, all relate to viewing something so without context it isn't immediately obvious which one V would be for.

seanbudd commented on source/gui/__init__.py:
```diff
@@ -514,31 +514,31 @@ def __init__(self, frame: MainFrame):
 				self.Bind(wx.EVT_MENU, frame.onInstallCommand, item)
 			# Translators: The label for the menu item to run the COM registration fix tool 
-			item = menu_tools.Append(wx.ID_ANY, _("Run COM Registration Fixing tool..."))
+			item = menu_tools.Append(wx.ID_ANY, _("&Run COM Registration Fixing tool..."))
```
this conflicts with `&Reload plugins`

```suggestion
			item = menu_tools.Append(wx.ID_ANY, _("Run COM Registration &Fixing tool..."))
```

CyrilleB79 replied:
> Please consider this carefully. For years now, people such as myself, have been giving, on mailing lists and in other support contexts, the instructions: **Press `NVDA+n`, `t`, `r` ...** to suggest that people attempt certain debugging tasks that include running the CRFT.

Thank you for this valuable feedback.

> Because of add-ons possibly conflicting with it, and `Reload plugins` also, the suggestion to press enter if necessary is often given, at least by me, to complete the instructions.

Pressing Enter is actually necessary, not because of add-ons but because there are two entries beginning with "R" and without explicit shortcut in the Tools menu: "Run COM Registration Fixing tool" and "Reload plugins".

Worse: if an add-on adds an item with explicit "R" shortcut in the "Tools" menu, this will mask both NVDA native items in current NVDA version (2023.2).
 
> Thus having a duplicate key for a minimally used function like `Reload plugins`, doesn't seem like much of a problem.

Personally, I prefer unique shortcut keys when possible since they are more efficient; no need to press enter nor to press various times the letter shortcut.

> However, if you change this key from `r` to `f`, no matter how much sense it makes, you will be invalidating all those support instructions. Or, at the least, calling their validity into question.
> 
> IMO, this key represents one of the top four uses of the tools menu--Add-on Store, Speech Viewer, Create Portable Copy, and the Com Reg Fixing Tool.
> 
> On the other hand, I do not recall anyone ever being told to reload plugins as part of a debugging step (or for any other reason, until very very recently--and in that case the `NVDA+Ctrl+F3` method was given).

I agree with changing the shortcut of "Reload plugins" instead given that there is already `NVDA+control+F3`. Since "p" is already taken by "Python console", "L" could be nice IMO.

seanbudd replied:
Once these are fixed it seems like https://github.com/nvaccess/nvda/issues/8913 can be closed too

Qchristensen replied:
I agree with the conversation and keeping the COM Registration key of R as per @XLTechie - Mostly commenting to note the earlier suggestion to make log viewer L (which would conflict with the suggestion to make reload plugins L)

seanbudd replied:
Apologies, I misread the issue

seanbudd reviewed (commented):
Thanks for fixing this up @CyrilleB79

CyrilleB79 committed:
Address review

seanbudd reviewed (dismissed):
Thanks @CyrilleB79

CyrilleB79 commented:
I have addressed all review comments in 98e6ba13757898bf6e82c96b4fb03b1429c77797. We end up with:
* View log  l
* Speech viewer  s
* Braille viewer  b
* Add-on store...  a
* Python console  p
* Create portable copy...  c
* Install NVDA...  i
* Reload plugins  e

View log, Add-on store and Reload plugins have their shortcut modified with respect to 2023.2. It should not cause any issue with previously given instructions since Add-on store is recent and other ones have associated NVDA gestures (NVDA+F1 and NVDA+control+F3).

seanbudd committed:
Merge remote-tracking branch 'origin/beta' into menuShortcuts

seanbudd committed:
update changes

tmthywynn8 commented:
Is it possible to change Is it possible to change "Run COM Registration Fixing tool..." to "T" and "Reload plugins" to "R"? Two reasons:

1."T' could be the mnemonic for "tool" and "R" for "reload". This makes way more sense than "E" not being a mnemonic for anything.
2. Users of the Remote add-on have been used to the "E" mnemonic for years, and while this is not part of NVDA core, making change for change's sake is not that great of a reason (see [this post][1] for a few disgruntled users).

[1]: https://dragonscave.space/@Brynify/111184847640354095

seanbudd commented:
Please open a new issue (feature request) thanks
